clients/src/main/java/org/apache/kafka/common/record/AbstractRecords.java

```diff
@@ -49,52 +50,67 @@ public boolean hasCompatibleMagic(byte magic) {
     }
 
     /**
-     * Convert this message set to use the specified message format.
+     * Convert this message set to a compatible magic format.
+     *
+     * @param toMaxMagic The maximum magic version to convert to. Entries with larger magic values
+     *                   will be converted to this magic; entries with equal or lower magic will not
+     *                   be converted at all.
      */
     @Override
-    public Records toMessageFormat(byte toMagic, TimestampType upconvertTimestampType) {
-        List<LogRecord> records = Utils.toList(records().iterator());
-        if (records.isEmpty()) {
+    public Records downconvert(byte toMaxMagic) {
```

Comment:
I think I understand why it is `toMaxMagic`, not `toMagic`, but I feel it is a bit awkward for readability. So I'd rather suggest we still name it `toMagic` and probably rename it in the future as it is private class anyways.